src/main/java/org/opensearch/ad/caching/CacheBuffer.java

```diff
@@ -324,21 +335,21 @@ public void maintenance() {
                     // put: not a problem as we are unlikely to maintain an entry that's not
                     // already in the cache
                     // remove method saves checkpoint as well
-                    remove(entityModelId);
-                } else {
-                    // we can have ConcurrentModificationException when serializing
-                    // and updating rcf model at the same time. To prevent this,
-                    // we need to have a deep copy of models or have a lock. Both
-                    // options are costly.
-                    // As we are gonna retry serializing either when the entity is
-                    // evicted out of cache or during the next maintenance period,
-                    // don't do anything when the exception happens.
-                    checkpointDao.write(modelState, entityModelId);
+                    removedStates.add(remove(entityModelId));
+                } else if (random.nextInt(6) == 0) {
+                    // checkpoint is relatively big compared to other queued requests
+                    // save checkpoints with 1/6 probability as we expect to save
+                    // all every 6 hours statistically
```

Comment:
Any reason to use random to decide whether we should save model checkpoints rather than some deterministic way? 
For this case: If some model checkpoints missing in index and we need to analyze why it's not saved, will be hard to tell if it's not saved as the random was not 6 or failed to save.
And as you said "checkpoint is relatively big", so I guess there will be some spike on metrics like disk write/CPU/memory when we save checkpoint? If we need to analyze the metrics, the random spikes may be very hard to explain.

Reply:
We will save a checkpoint when
1) removing the model from cache.
2) cold start
3) no complete model only a few samples.  If we don't save new samples, we will never be able to have enough samples for a trained mode.
4) periodically save in case of exceptions.

This branch is doing 4). Previously, I will do it every hour for all in-cache models.  Consider we are moving to 1M entities, this will bring the cluster in a heavy payload every hour.  That's why I am doing it randomly (expected 6 hours for each checkpoint statistically).  

I am doing it random since maintaining a state of which one has been saved and which one hasn't are not cheap.  Also, the models in the cache can be dynamically changing.  Will have to maintain the state in the removing logic.  Random is a lazy way to deal with this as it is stateless and statistically sound.  

We will have a checkpoint, just that if a checkpoint does not fall into the 6-hour bucket in a particular scenario, the model is stale (i.e., we don't recover from the freshest model in disaster.).

Yes, this change is mostly due to performance and easy maintenance.  

Yes, the solution has flaws.  It is a tradeoff.  Any other suggestions?

Reply:
Thanks, look good as tradeoff